DOC: Update README with proposed schema conventions (#1112)
bids-standard/bids-specification: src/schema/README.md

the entity tables, entity definitions, filename templates, and metadata tables.
Any changes to the specification should be mirrored in the schema.

> 

Reviewer: This file is getting long enough that a ToC might be warranted.

Author: Thought about it. Would be its own thing to maintain. Note that GitHub has something built-in when viewing the file, though sadly not the diff:

![image](https://user-images.githubusercontent.com/83442/191313154-271f88a9-7b14-4552-bdb9-d50636cebac4.png)

Reviewer: I'm waffling on whether the very specific sections of each of the top level namespaces should be split into their own README in that subdirectory.

Author: Not a bad idea. Would rather not do it right now.

## Organization and syntax

At the time of this writing, the schema has the following file layout:

```plain
src/schema
├── BIDS_VERSION
├── meta
│   └── context.yaml
├── objects
│   ├── associated_data.yaml
│   ├── ...
│   └── top_level_files.yaml
├── rules
│   ├── associated_data.yaml
│   ├── checks
│   │   ├── asl.yaml
│   │   ├── ...
│   │   └── mri.yaml
│   ├── ...
│   └── top_level_files.yaml
└── SCHEMA_VERSION
```

The top-level organization includes `objects`, where terms are defined;
`rules`, where constraints (such as valid filenames or required metadata fields)
are defined;
and `meta`, where definitions useful for interpreting the schema are defined.

> Each file is a YAML structure, most often an *object*.

Reviewer: ```suggestion
Each file is made up of YAML data, most often an *object*.
```

and `rules.checks.mri.PhasePartUnits.issue` refers to the object:

```JSON
{
  "code": "PHASE_UNITS",
  "message": "Phase images (with the `part-phase` [...]\n\"rad\" or \"arbitrary\".\n",
  "level": "error"
}
```

These qualified names may be used in this README, as well as in *references* and
*expressions*.

### Description formatting

Many objects throughout the schema have `description` fields, which will typically be
rendered somewhere in the specification. Because the specification is written in
[Markdown](https://en.wikipedia.org/wiki/Markdown), these description fields may also
contain Markdown, including links to other locations in the specification.

Because the same description may be used in multiple locations, a mechanism is needed
to ensure that the correct path is discovered to render the description in each location.
To do this, the path should follow the form `SPEC_ROOT/path/within/source.md#anchor`.
For example, to link to the
[Definitions](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#definitions)
section of
[Common principles](https://bids-specification.readthedocs.io/en/stable/02-common-principles.html),

> use the path `SPEC_ROOT/02-common-principles.md#definitions`, e.g.,

Author: ```suggestion
use the path `SPEC_ROOT/02-common-principles.md#definitions`:
```

arrays or objects:

| Field                  | Description                                                                                   |
| ---------------------- | --------------------------------------------------------------------------------------------- |
| `anyOf`                | A list of constraints, any of which could apply                                               |
| `items`                | The array described contains values whose types are constrained                               |
| `properties`           | The object described has a given set of fields; the values of these fields may be constrained |
| `additionalProperties` | The object described has constraints on its values, but not the names                         |

### On re-used objects with different definitions

In a few cases, two objects with the same name appear multiple times in the specification.
When this happens, it is preferred to find a common definition, and clarify it in the rules (see below).
However, in some cases, the object description and permissible values differ, and it needs to be defined
as two separate objects.

> The following conventions are used:

Author: Done

    | `description`  | Term definition     |

-   `objects.top_level_files`
    | Field          | Description         |
    | -------------- | ------------------- |
    | `display_name` | Human-friendly name |
    | `description`  | Term definition     |

## Rule files

The `rules.*` namespace contains most of the validatable content of the schema,
apart from value constraints that can be encoded in `objects`.

There are several types of rule, and this section is subject to reconsolidation as
patterns are found.

### Core concepts

Core concepts are [expressions](#expressions) (defined above), requirement levels and issues.

#### Requirement levels and severity

BIDS follows RFC 2119 and has three requirement levels: OPTIONAL, RECOMMENDED and REQUIRED.
In the schema, we use `optional`, `recommended` and `required`. A rule interpreter (validator)
is expected to treat missing REQUIRED data/metadata as an error, missing RECOMMENDED
data/metadata as a warning, and silently pass over missing OPTIONAL data.

BIDS also defines a level `DEPRECATED`, rendered in the schema as `deprecated`, and corresponding
to a warning if the data/metadata is present.

> 

Author: It does not exist so far. I'd like to avoid it until it becomes definitely needed.

> 

Reviewer 2: Can't that case be covered with a `!(ImProhibited in blah)`?

> 

Author: That would be a check, not a requirement level. So yes, it can be done so far. Maybe there will be a case where we'd have to bend over backwards to avoid it and it makes more sense to go ahead. Haven't seen it yet.

> 

Reviewer 2: @rwblair maintains the validator, so I could go either way on this :)

> 

Reviewer: There's a pr some where with prohibited as a requirement level, I think its useful for the mutual exclusion cases.

#### Issues

Issues are messages intended to be communicated to a dataset curator to indicate an issue
with their dataset. They have a code and severity as well:

| Field     | Description                                    |
| --------- | ---------------------------------------------- |
| `code`    | Issue identifier, such as `EVENTS_TSV_MISSING` |
| `level`   | Issue severity (`warning` or `error`)          |
| `message` | Message for display to a user                  |

A level of `warning` corresponds to a rule in the specification that is RECOMMENDED, while a
level of `error` corresponds to a rule that is REQUIRED.

In some cases, an issue is contained next to a `level: required` or `level: recommended` as part
of a larger rule. In these cases, the `level` field should be omitted from the issue to avoid
duplication or conflict.

### Filename construction rules

A significant portion of BIDS is devoted to the naming of files, and almost all file names consist

> of entities, a suffix, an extension, and a data type. Exceptions will be noted below.

Author: This documents the current state.

I would like to rename `rules.datatypes` to `rules.filenames` and include `rules.associated_data` and `rules.top_level_files` inside this.

while files in the second group may not.
Also, when files in the second group have the `acq` entity, the associated value MUST be `crosstalk`.

### `sidecars/*.yaml`

Sidecar files introduce the idea of "selectors" to determine whether a set of fields is added
to the sidecar schema that will be applied to the sidecar.

> From here, the `fields` property lists will be accumulated to set requirement levels.

Reviewer 3: This is confusingly worded to me. Perhaps

```suggestion
From here, the metadata fields listed in the `fields` property will be accumulated across dictionaries to set requirement levels after evaluating all selectors.
```

That's not really any better, actually.

      Task scans should have a corresponding events.tsv file.
      If this is a resting state scan you can ignore this warning or rename the task to include the word "rest".
    level: warning # could be an error with the proper selectors, I think
  selectors:
    - '"task" in entities'
    - '!matches(entities.task, "rest")'
    - suffix != "events"
  checks:
    - '"events" in associations'
```

### Ordering rules

-   `rules.entities` - This file contains a list of keys into `objects.entities` and
    simply defines the order in which entities, when present, MUST appear in filenames

-   `rules.common_principles` - This file contains a list of terms that appear in `objects.common_principles`
    that determines the order they appear in the specification

### One-off rules

-   `rules.modalities` - The keys in this file are the modalities, the values objects with the following field:
    | Field       | Description                           |
    | ----------- | ------------------------------------- |
    | `datatypes` | List of datatypes mapping to modality |

-   `rules.dataset_metadata` - These are similar to `rules.sidecars.*`, for JSON files at the root level.

>     This is likely to go away in favor of other approaches.

Author: I do not like this, so I didn't bother fleshing it out. I think what we want is to fully define these documents inside the `objects.*` namespace, and find some way in the rules to say something like "a file named `dataset_description.json` contains the object `objects.metadata.DatasetDescription`".
